api/METRICS.md

```diff
@@ -86,21 +86,186 @@ group_max_from_summary(summary)
+
+**Proposal.** The following proposal is an extension of the status quo; no breaking changes are introduced:
+
+* `constraints`: in addition to `"demographic_parity"` and `"equalized_odds"`, also allow:
+
+  * `"<metric>_parity"`: for all _a_,<br>
+    _metric_(_a_) = _metric_(\*).
+
+* `objective` is a string of the form:
+
+  * `"<metric>"`: <br>
+    goal is then to maximize _metric_(\*) subject to constraints
+
+### fairlearn.reductions
+
+We support two algorithms: `ExponentiatedGradient` and `GridSearch`. They both represent `constraints` and `objective` via objects of type `Moment`.
+
+**Status quo:**
+
+* In both cases, `objective` is automatically inferred from the provided `constraints`.
+
+* `ExponentiatedGradient(estimator, constraints, eps=`&epsilon;`)` considers constraints that are specified jointly by the provided `Moment` object and the numeric value &epsilon; as follows:
+
+  * `DemographicParity()`: for all _a_, <br>
+    |_selection_rate_(_a_) - _selection_rate_(\*)| &le; &epsilon;.
+
+  * `DemographicParity(ratio=`_r_`)`: for all _a_, <br>
+    _r_ &sdot; _selection_rate_(_a_) - _selection_rate_(\*) &le; &epsilon;, <br>
+    _r_ &sdot; _selection_rate_(\*) - _selection_rate_(_a_) &le; &epsilon;.
+
+  * `TruePositiveRateDifference()`: for all _a_, <br>
+    |_true_positive_rate_(_a_) - _true_positive_rate_(\*)| &le; &epsilon;.
+
+  * `TruePositiveRateDifference(ratio=`_r_`)`: analogous
+
+  * `EqualizedOdds()`: for all _a_, <br>
+    |_true_positive_rate_(_a_) - _true_positive_rate_(\*)| &le; &epsilon;, <br>
+    |_false_positive_rate_(_a_) - _false_positive_rate_(\*)| &le; &epsilon;.
+
+  * `EqualizedOdds(ratio=`_r_`)`: analogous
+
+  * `ErrorRateRatio()`: for all _a_, <br>
+    |_error_rate_(_a_) - _error_rate_(\*)| &le; &epsilon;.
+
+  * `ErrorRateDifference(ratio=`_r_`)`: analogous
+
+  * all of the above constraints are descendants of `ConditionalSelectionRate`
+
+  * the `objective` for all of the above constraints is `ErrorRate()`
+
+* `GridSearch(estimator, constraints)` considers constraints represented by the provided `Moment` object. The behavior of the `GridSearch` algorithm does not depend on the value of the right-hand side of constraints, so it is not provided to the constructor. In addition to the `Moment` objects above, `GridSearch` also supports the following:
+
+  * `GroupLossMoment(<loss>)`: for all _a_, <br>
+    _loss_(_a_) &le; &zeta;
+
+    * where `<loss>` is a _loss evaluation_ object; it supports an interface that takes `y_true` and `y_pred` as the input and returns the vector of losses evaluated on individual examples
+
+    * the `objective` for `GroupLossMoment(<loss>)` is `AverageLossMoment(<loss>)`
+
+  * both `GroupLossMoment` and `AverageLossMoment` are descendants of `ConditionalLossMoment`
+
+**Proposal.** The proposal introduces some breaking changes:
+
+* `constraints`:
+
+  * `<Metric>Parity(difference_bound=`&epsilon;`)`: for all _a_, <br>

```

Comment:
Python doesn't do overloading with various class constructor definitions, but rather with operator overloading, so this would look like this
```python
<Metric>Parity(difference_bound, ratio_bound, ratio_bound_slack)
```
Is that what you had in mind?

Reply:
... and then a pile of logic to make sure that you've given it a consistent set of arguments.... :-/

Reply:
Actually, what I had in mind was more something like
```python
<Metric>Parity(*, difference_bound=None, ratio_bound=None, ratio_bound_slack=0.0)
```

Reply:
... and that's what Andra and Lauren are using right now, so I'll close this.